src/LanguageServer/Protocol/Features/Options/SolutionAnalyzerConfigOptionsUpdater.cs

```diff
@@ -0,0 +1,273 @@
+                {
+                    return;
+                }
+
+                UpdateSolution(
+                    transformation: solution => InitializeLanguages(solution, _initializedLanguages),
+                    onAfterUpdate: solution => _initializedLanguages = solution.SolutionState.ProjectCountByLanguage);
+            }
+            catch (Exception e) when (FatalError.ReportAndPropagate(e, ErrorSeverity.Diagnostic))
+            {
+                throw ExceptionUtilities.Unreachable();
+            }
+        }
+
+        public void GlobalOptionsChanged(object? sender, OptionChangedEventArgs args)
+        {
+            try
+            {
+                if (!args.ChangedOptions.Any(static o => o.key.Option.Definition.IsEditorConfigOption))
+                {
+                    return;
+                }
+
+                UpdateSolution(
+                    transformation: solution => UpdateOptions(solution, args, _initializedLanguages),
+                    onAfterUpdate: null);
+            }
+            catch (Exception e) when (FatalError.ReportAndPropagate(e, ErrorSeverity.Diagnostic))
+            {
+                throw ExceptionUtilities.Unreachable();
+            }
+        }
+
+        private void UpdateSolution(Func<Solution, Solution> transformation, Action<Solution>? onAfterUpdate)
+        {
+            var lockTaken = false;
+            try
+            {
+                // If another update is in progress wait until it completes.
+                Monitor.Enter(_updateLock, ref lockTaken);
+
+                workspace.SetCurrentSolution(
+                    transformation,
+                    changeKind: WorkspaceChangeKind.SolutionChanged,
+                    onAfterUpdate: (_, newSolution) =>
+                    {
+                        onAfterUpdate?.Invoke(newSolution);
+
+                        // unlock before workspace events are triggered:
+                        if (lockTaken)
+                        {
+                            Monitor.Exit(_updateLock);
+                            lockTaken = false;
+                        }
+                    });
+            }
+            finally
+            {
+                if (lockTaken)
+                {
+                    Monitor.Exit(_updateLock);
+                }
+            }
+        }
+
+        private Solution InitializeLanguages(Solution solution, ImmutableDictionary<string, int> initializedLanguages)
+        {
+            var oldFallbackOptions = solution.FallbackAnalyzerOptions;
+            var newFallbackOptions = oldFallbackOptions;
+
+            // Clear out languages that are no longer present in the solution.
+            // If we didn't, the workspace might clear the solution (which removes the fallback options)
+            // and we would never re-initialize them from global options.
+            foreach (var (language, _) in initializedLanguages)
```

Comment:
do we ever use the project count?  or is this really just a question of how many different languages we have.  if so, consider having the workspace expose *that* (although internally it could still have hte count for an effective internal ref counting system).

Reply:
We only need the languages, not the count. However, exposing just the keys would require extra allocations since ImmutableDictionary.Keys creates a new enumerator.

Reply:
We would also need to add ContainsLanguage to replace ContainsKey.

Reply:
> However, exposing just the keys would require extra allocations since ImmutableDictionary.Keys creates a new enumerator.

i'm genuinely ok with that.  It feels pretty unlikely that that tiny allocation will matter.  let's go with that first, and then optimize to be no-alloc if data shows any sort of problem.

Reply:
Will follow up with a generalized LanguageMap<T>